File: app/src/main/java/com/example/david/lists/view/itemlist/IItemViewContract.java
package com.example.david.lists.view.itemlist;

import com.example.david.lists.data.datamodel.Item;

import java.util.List;

public interface IItemViewContract {
    interface View {
        void openAddDialog(String userListId, int position);

        void openEditDialog(Item item);

        void submitList(List<Item> viewData);

        void notifyUserOfDeletion(String message);

        void setStateDisplayList();

        void setStateLoading();

        void setStateError(String message);

        void showMessage(String message);

        void finishView();
    }

    interface Logic {
        void onStart();

        void add();

        void edit(int position);

        void dragging(int fromPosition, int toPosition, IItemViewContract.Adapter adapter);

        void movedPermanently(int newPosition);

        void delete(int position, IItemViewContract.Adapter adapter);

        void undoRecentDeletion(IItemViewContract.Adapter adapter);

        void deletionNotificationTimedOut();

        void onDestroy();
    }

    interface ViewModel {
        String getUserListId();

        void setViewData(List<Item> items);

        List<Item> getViewData();

        void setTempPosition(int position);

        List<Item> getTempList();

        int getTempPosition();

        String getMsgListDeleted(String title);

        String getMsgItemDeleted();

        String getErrorMsg();

        String getErrorMsgEmptyList();

        String getErrorMsgInvalidUndo();
    }

    interface Adapter {
        void submitList(List<Item> list);

        void move(int fromPosition, int toPosition);

        void remove(int position);

        void reAdd(int position, Item item);
    }
}

File: app/src/main/java/com/precopia/david/lists/view/itemlist/IItemViewContract.kt
package com.precopia.david.lists.view.itemlist

import androidx.lifecycle.LiveData
import com.precopia.domain.datamodel.Item

interface IItemViewContract {
    interface View

    interface Logic {
        fun onEvent(even: LogicEvents)

        fun observe(): LiveData<ViewEvents>
    }

    interface ViewModel {
        val userListId: String

        var viewData: MutableList<Item>

        val tempList: MutableList<Item>

        var tempPosition: Int

        val msgDeletion: String

        fun getMsgListDeleted(title: String): String

        val errorMsg: String

        val errorMsgEmptyList: String

        val errorMsgInvalidUndo: String
    }

    interface Adapter {
        fun setData(list: List<Item>)

        fun move(fromPosition: Int, toPosition: Int)

        fun remove(position: Int)

        fun reAdd(position: Int, item: Item)
    }


    sealed class ViewEvents {
        data class OpenAddDialog(val userListId: String, val position: Int) : ViewEvents()
        data class OpenEditDialog(val item: Item) : ViewEvents()
        data class SetViewData(val viewData: List<Item>) : ViewEvents()
        data class NotifyUserOfDeletion(val message: String) : ViewEvents()
        object SetStateDisplayList : ViewEvents()
        object SetStateLoading : ViewEvents()
        data class SetStateError(val message: String) : ViewEvents()
        data class ShowMessage(val message: String) : ViewEvents()
        object FinishView : ViewEvents()
    }

    sealed class LogicEvents {
        object OnStart : LogicEvents()
        object Add : LogicEvents()
        data class Edit(val position: Int) : LogicEvents()
        data class Dragging(val fromPosition: Int, val toPosition: Int, val adapter: Adapter) : LogicEvents()
        data class MovedPermanently(val newPosition: Int) : LogicEvents()
        data class Delete(val position: Int, val adapter: Adapter) : LogicEvents()
        data class UndoRecentDeletion(val adapter: Adapter) : LogicEvents()
        object DeletionNotificationTimedOut : LogicEvents()
    }
}
